src/telemetry.ts

```diff
@@ -16,41 +18,100 @@ interface IPackageInfo {
     aiKey: string;
 }
 
-let telemetryReporter: TelemetryReporter | null;
+export class ExperimentationTelemetry implements IExperimentationTelemetry {
+  private sharedProperties: Record<string, string> = {};
 
-export function activate(): void {
-    try {
-        telemetryReporter = createReporter();
-    } catch (e) {
-        // can't really do much about this
-    }
+  constructor(private baseReporter: TelemetryReporter) { }
+
+  sendTelemetryEvent(eventName: string, properties?: Record<string, string>, measurements?: Record<string, number>): void {
+      this.baseReporter.sendTelemetryEvent(
+          eventName,
+          {
+              ...this.sharedProperties,
+              ...properties
+          },
+          measurements
+      );
+  }
+
+  sendTelemetryErrorEvent(eventName: string, properties?: Record<string, string>, _measurements?: Record<string, number>): void {
+      this.baseReporter.sendTelemetryErrorEvent(eventName, {
+          ...this.sharedProperties,
+          ...properties
+      });
+  }
+
+  setSharedProperty(name: string, value: string): void {
+      this.sharedProperties[name] = value;
+  }
+
+  postEvent(eventName: string, props: Map<string, string>): void {
+      const event: Record<string, string> = {};
+      for (const [key, value] of props) {
+          event[key] = value;
+      }
+      this.sendTelemetryEvent(eventName, event);
+  }
+
+  dispose(): Promise<any> {
+      return this.baseReporter.dispose();
+  }
 }
 
-export async function deactivate(): Promise<void> {
-    if (telemetryReporter) {
-        await telemetryReporter.dispose();
+let initializationPromise: Promise<IExperimentationService> | undefined;
+let experimentationTelemetry: ExperimentationTelemetry | undefined;
+
+export function activate(extensionContext: vscode.ExtensionContext): void {
+  try {
+    if (extensionContext) {
+      const packageInfo: IPackageInfo = getPackageInfo();
+      if (packageInfo) {
+        const targetPopulation: TargetPopulation = TargetPopulation.Public;
+        experimentationTelemetry = new ExperimentationTelemetry(new TelemetryReporter(packageInfo.name, packageInfo.version, appInsightsKey));
+        initializationPromise = getExperimentationServiceAsync(packageInfo.name, packageInfo.version, targetPopulation, experimentationTelemetry, extensionContext.globalState);
+      }
     }
+  } catch (e) {
+    // Handle error with a try/catch, but do nothing for errors.
+  }
 }
 
-export function logEvent(eventName: string, properties?: Properties, measures?: Measures): void {
-    if (telemetryReporter) {
-        telemetryReporter.sendTelemetryEvent(eventName, properties, measures);
-    }
+export async function getExperimentationService(): Promise<IExperimentationService | undefined> {
```

Comment:
This is potentially undefined

Reply:
I'm not sure what else to do here, were you recommending I change something?

Reply:
Update the annotation to `Promise<IExperimentationService | undefined> | undefined` or call `getExperimentationServiceAsync` and assign `initializationPromise` here before returning it back.

Reply:
If you assign it here, you'll need to update `activate()`